aspnetcore/blazor/security/server/qrcodes-for-authenticator-apps.md

```diff
@@ -0,0 +1,143 @@
+
+For more information, see <xref:security/authentication/scaffold-identity>. For guidance on using the .NET CLI instead of Visual Studio, see <xref:fundamentals/tools/dotnet-aspnet-codegenerator>.
+
+## Adding QR codes to the 2FA configuration page
+
+These instructions use [Shim Sangmin](https://hogangnono.com)'s [qrcode.js: Cross-browser QRCode generator for JavaScript](https://davidshimjs.github.io/qrcodejs/) ([`davidshimjs/qrcodejs` GitHub repository](https://github.com/davidshimjs/qrcodejs)).
+
+Download the [`qrcode.min.js`](https://davidshimjs.github.io/qrcodejs/) library to the `wwwroot` folder of the solution's server project. The library has no dependencies.
+
+In the `App` component (`Components/App.razor`), place a library script reference after [Blazor's `<script>` tag](xref:blazor/project-structure#location-of-the-blazor-script):
+
+```razor
+<script src="qrcode.min.js"></script>
+```
+
+The `EnableAuthenticator` component, which is part of the QR code system in the app and displays the QR code to users, adopts static server-side rendering (static SSR) with enhanced navigation. Therefore, normal scripts can't execute when the component loads or updates under enhanced navigation. Extra steps are required to trigger the QR code to load in the UI when the page is loaded. To accomplish loading the QR code, the approach explained in <xref:blazor/js-interop/ssr> is adopted.
+
+Add the following [JavaScript initializer](xref:blazor/fundamentals/startup#javascript-initializers) to the server project's `wwwroot` folder. The `{NAME}` placeholder must be the name of the app's assembly in order for Blazor to locate and load the file automatically. If the server app's assembly name is `BlazorSample`, the file is named `BlazorSample.lib.module.js`.
+
+`wwwroot/{NAME}.lib.module.js`:
+
+[!INCLUDE[](~/blazor/includes/js-interop/blazor-page-script.md)]
+
+Add the following shared `PageScript` component to the server app.
+
+`Components/PageScript.razor`:
+
+```razor
+<page-script src="@Src"></page-script>
+
+@code {
+    [Parameter]
+    [EditorRequired]
+    public string Src { get; set; } = default!;
+}
+```
+
+Add the following [collocated JS file](xref:blazor/js-interop/index#load-a-script-from-an-external-javascript-file-js-collocated-with-a-component) for the `EnableAuthenticator` component, which is located at `Components/Account/Pages/Manage/EnableAuthenticator.razor`. The `onLoad` function creates the QR code with Sangmin's `qrcode.js` library using the QR code URI produced by the `GenerateQrCodeUri` method in the component's `@code` block.
+
+`Components/Account/Pages/Manage/EnableAuthenticator.razor.js`:
+
+```javascript
+export function onLoad() {
+  const uri = document.getElementById('qrCodeData').getAttribute('data-url');
+  new QRCode(document.getElementById('qrCode'), uri);
+}
+
+export function onUpdate() { }
+
+export function onDispose() { }
+```
```

Comment:
Nit: It's not necessary to include the `onUpdate()` and `onDispose()` callbacks if they don't do anything :) The code will still work.